A lot of dotty-specific changes to stdlib. [ci: last-only]

DarkDimius opened the pull request:
There are going to be a lot more. We need to have some procedure to backport those fixes to scala/scala.

@adriaanm Am I right in my assumption that it would be simpler for you if we have a single bulk PR for stdlib changes? We're not done yet and _I do not propose this PR for review_, but to discuss intention of merging changes to scala/scala and to showcase what changes are needed.

DarkDimius committed:
Remove procedure syntax from standard library.

Is it planned for deprecation.

DarkDimius committed:
Make return types of implicits explicit.

Explicit return types of implicits are required in Dotty.

DarkDimius committed:
Remove procedure syntax in LinearSeqOptimized.

DarkDimius committed:
TraversableOnce: make return types explicit

DarkDimius committed:
Faster compare for BooleanOrdering: no tupling

DarkDimius committed:
Faster compare for OptionOrdering: no tupling

DarkDimius committed:
Make math classes compile under Dotty.

In dotty you can't have an inner class overriding.

DarkDimius committed:
Modify partial function so that Dotty compiles it fine.

x has type Any in dotty so f(x) does not type check.

DarkDimius committed:
@odersky: Dotty can't find scala.reflect.Manifest object.

DarkDimius committed:
Growable: use Unit syntax

DarkDimius committed:
Growable: make type checkable by Dotty.

Dotty is more restrictive in case of this pattern match.

DarkDimius committed:
Queue.scala: adapt pattern to be type checkable by Dotty

Dotty is more conservative in covariant types in patterns.

DarkDimius committed:
mutable.Builder: use unit syntax.

DarkDimius committed:
immutable.Vector: use unit-syntax.

DarkDimius committed:
mutable.GrowingBuilder: Use unit syntax

DarkDimius committed:
immutable.RedBlackTree: use unit-syntax.

DarkDimius committed:
mutable.Builder: Dotty does not refine types of override members

It needs to be asked explicitly to do it.

Ichoran commented on src/library/scala/collection/mutable/ListBuffer.scala:
```diff
@@ -69,7 +69,7 @@ final class ListBuffer[A]
   protected def underlying: List[A] = start
 
-  private def writeObject(out: ObjectOutputStream) {
+  private def writeObject(out: ObjectOutputStream) = {
```
Need explicit annotation of Unit here (and in a bunch of other places).

Ichoran commented on src/library/scala/collection/TraversableOnce.scala:
```diff
@@ -402,8 +402,8 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
 object TraversableOnce {
-  implicit def alternateImplicit[A](trav: TraversableOnce[A]) = new ForceImplicitAmbiguity
-  implicit def flattenTraversableOnce[A, CC[_]](travs: TraversableOnce[CC[A]])(implicit ev: CC[A] => TraversableOnce[A]) =
+  implicit def alternateImplicit[A](trav: TraversableOnce[A]): ForceImplicitAmbiguity = new ForceImplicitAmbiguity
```
It's the return of Java boilerplate!  Are we really advocating this as best practice (explicit return type when you are literally just calling the constructor of that very type)?

adriaanm replied:
For implicits, yes, for some TBD value of "yes". Currently, type inference and implicit search cycles can cause very hard to understand implicit search ambiguities. If we can come up with some easy to understand scheme that softens this requirement, great! Otherwise, I'm on board with the change in principle.

adriaanm commented:
Let's target 2.12.x for this, please. Also, could you include `[ci: last-only]` in the title of your 2.12 PR, while we're experimenting, to [avoid having the CI build every commit](https://github.com/scala/scabot/blob/master/README.md#pr-title-modifiers) until the PR is ready?

adriaanm commented on src/library/scala/Option.scala:
```diff
@@ -253,8 +253,8 @@ sealed abstract class Option[+A] extends Product with Serializable {
    */
-  @inline final def foreach[U](f: A => U) {
-    if (!isEmpty) f(this.get)
+  final def foreach[U](f: A => U): Unit = {
```
I don't think we should remove `@inline` here.

adriaanm commented on src/library/scala/math/Ordering.scala:
```diff
@@ -224,11 +224,7 @@ object Ordering extends LowPriorityOrderingImplicits {
-      case (true, false) => 1
-      case _ => 0
-    }
+    def compare(x: Boolean, y: Boolean) = if(x) { if (y) 0 else 1} else {if (y) -1 else 0}
```
why is this change needed?

lrytz replied:
territory for the new optimizer! i'll be working on this hopefully next week.

sjrd replied:
+1! Scala.js can do it:

``` javascript
ScalaJS.s.s_math_Ordering$BooleanOrdering$class__compare__s_math_Ordering$BooleanOrdering__Z__Z__I = (function($$this, x, y) {
  if ((!x)) {
    if (y) {
      return (-1)
    }
  };
  if (x) {
    if ((!y)) {
      return 1
    }
  };
  return 0
});
```

lrytz replied:
🏆 :-)

DarkDimius replied:
>  why is this change needed?

It's not needed. It's just that I saw how inefficient the code was and I didn't resist the temptation to rewrite it. I'm fine with dropping this change if it's not welcome.

adriaanm replied:
In big changes like this, we like to separate concerns. The PR  should be restricted to changes that are crucial, and anything non-obvious must be motivated. Let's discuss whether to split up along module boundaries or along the kind of change. I think it makes most sense to have one commit per kind of change (procedure syntax, explicit result types for implicits, avoid name clashes for inner classes). I would also like to see a bytecode diff. We have automated this: https://github.com/retronym/libscala/tree/master/git-java (git diff invokes a custom command to generate only the essential diff). Here's an asm-based javap replacement that sorts by method name: https://github.com/adriaanm/scala/commit/afd15bb291b81dff4df4c22c384a4ab2c4714fad

lrytz replied:
For the record, scalac can now also do it :) will add it as a test case. cfr-decompiler shows

```
    public int compare(boolean x, boolean y) {
        int n = false == x && true == y ? -1 : (true == x && false == y ? 1 : 0);
        return n;
    }
```

sjrd replied:
Nice!

adriaanm replied:
Indeed!

DarkDimius replied:
:+1:

soc commented:
As @adriaanm, please target 2.12.x. I'm happy to review!

som-snytt commented:
You can use the code rewriting tool to remove procedure syntax.

Oh wait, DarkDimius _is_ the code rewriting tool.

DarkDimius committed:
Make return types of implicits explicit.

DarkDimius committed:
Procedure syntax.

DarkDimius committed:
Add type tests in HashMap and SortedSeqLike

DarkDimius committed:
IntMap: make return type of implicit explicit.

DarkDimius committed:
Range: make return types of implicits explicit.

DarkDimius committed:
Fix procedure syntax here and there

odersky committed:
Add missing type parameters to Predef's TupleXToZipped.Ops

odersky committed:
Avoid class overriding clash in SortedMap

odersky committed:
Add explicit type for view

Otherwise the inherited type is picked, which is too weak in this case.

odersky committed:
Explicitly type-annotate IterableLie and TraversableLike # view

Here's the situation in the current stdlib:

In TraversableLike:

    def view = new TraversableView[A, Repr] {
      protected lazy val underlying = self.repr
      override def foreach[U](f: A => U) = self foreach f
    }

in IterableLike:

    override def view = new IterableView[A, Repr] {
      protected lazy val underlying = self.repr
      override def iterator = self.iterator
    }

In Dotty, we have to annotate the type in IterableLike because otherwise the
inherited type from TraversableLike is chosen. So the IterableLike view becomes:

    override def view: IterableView[A, Repr] = new IterableView[A, Repr] {
      protected lazy val underlying = self.repr
      override def iterator = self.iterator
    }

But now we also have to annotate the view in TraversableLike because otherwise
its inferred type is

    TraversableView[A, Repr] { val underlying: Repr }

which is not a supertype of `IterableView[A, Repr]`. Or else we add the `underlying`
refinement to `IterableLike#view`. But the intention is that `underlying` should
not be part of the type.

smarter committed:
Make immutable.Set#parCombiner object-protected

This matches every other override of parCombiner in the library and is
necessary to avoid the following variance issue detected by dotty:

src/library/scala/collection/immutable/Set.scala:41: error: covariant
type A occurs in contravariant position in type
=> scala.collection.parallel.Combiner[A,
scala.collection.parallel.immutable.ParSet[A]] of method parCombiner

odersky committed:
Add missing implicit result type

odersky committed:
Avoid structural type lambda in definition of Set.

odersky committed:
Merge branch 'dotty-library' of https://github.com/DarkDimius/scala into dotty-library

odersky committed:
Refactoring of TrieIterator to do without early definitions.

odersky committed:
Rewrite of ...View classes to not use existential types.

odersky committed:
Fix ViewLike tests to make them Dotty compatible.

Quite a lot of work needed there.

lrytz commented:
closing this, as it targets the wrong branch anyway
